src/MSALWrapper/AuthFlow/DeviceCode.cs

```diff
@@ -0,0 +1,163 @@
+        private readonly IList<Exception> errors;
+        private IPCAWrapper pcaWrapper;
+
+        #region Public configurable properties
+
+        /// <summary>
+        /// The silent auth timeout.
+        /// </summary>
+        private TimeSpan silentAuthTimeout = TimeSpan.FromMinutes(5);
+
+        /// <summary>
+        /// The device code flow timeout.
+        /// </summary>
+        private TimeSpan deviceCodeFlowTimeout = TimeSpan.FromMinutes(15);
+        #endregion
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeviceCode"/> class.
+        /// </summary>
+        /// <param name="logger">The logger.</param>
+        /// <param name="clientId">The client id.</param>
+        /// <param name="tenantId">The tenant id.</param>
+        /// <param name="scopes">The scopes.</param>
+        /// <param name="osxKeyChainSuffix">The osx key chain suffix.</param>
+        /// <param name="preferredDomain">The preferred domain.</param>
+        /// <param name="pcaWrapper">Optional: IPCAWrapper to use.</param>
+        /// <param name="promptHint">The customized header text in account picker for WAM prompts.</param>
+        public DeviceCode(ILogger logger, Guid clientId, Guid tenantId, IEnumerable<string> scopes, string osxKeyChainSuffix = null, string preferredDomain = null, IPCAWrapper pcaWrapper = null, string promptHint = null)
+        {
+            this.errors = new List<Exception>();
+            this.logger = logger;
+            this.scopes = scopes;
+            this.preferredDomain = preferredDomain;
+            this.promptHint = promptHint;
+            this.pcaWrapper = pcaWrapper ?? this.BuildPCAWrapper(logger, clientId, tenantId, osxKeyChainSuffix);
+        }
+
+        /// <summary>
+        /// Gets the token for a resource.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> of <see cref="TokenResult"/>.</returns>
+        public async Task<AuthFlowResult> GetTokenAsync()
+        {
+            IAccount account = await this.pcaWrapper.TryToGetCachedAccountAsync(this.preferredDomain)
+                ?? null;
+            this.logger.LogDebug($"GetTokenNormalFlowAsync: Using account '{account.Username}'");
+
+            try
+            {
+                try
+                {
+                    var tokenResult = await TaskExecutor.CompleteWithin(
+                        this.logger,
+                        this.silentAuthTimeout,
+                        "Get Token Silent",
+                        (cancellationToken) => this.pcaWrapper.GetTokenSilentAsync(
+                            this.scopes,
+                            account,
+                            cancellationToken),
+                        this.errors)
+                        .ConfigureAwait(false);
+                    tokenResult.SetAuthenticationType(AuthType.Silent);
+
+                    return new AuthFlowResult(tokenResult, this.errors);

```

Comment:
Could you not also just let this be handled by the final `return` by declaring `tokenResult` outside the scope of this series of `try` blocks. It would have to be `null` initially, but that would just match the final `return` we have here.

```csharp
var tokenResult = null;
```

Then, in this place you'd just set
```csharp
tokenResult = await TaskExecutor.CompleteWithin(...);
```

If you did it this way you've have a single `return` statement for this whole method, which I find helps make things easier to keep track of.

```csharp
return new AuthFlowResult(tokenResult, this.errors);
```

Reply:
The code is effectively the same. Let's be consistent for now. It's a minor modification we can return to later in another refactor if we want to.